pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java

```diff
@@ -356,22 +358,37 @@ public boolean removeKeyCipher(String keyName) {
      * 
      * @return encryptedData if success
      */
-    public synchronized ByteBuf encrypt(ConcurrentOpenHashSet<String> encKeys, MessageMetadata.Builder msgMetadata, ByteBuf payload)
-            throws PulsarClientException {
+    public synchronized ByteBuf encrypt(ConcurrentOpenHashSet<String> encKeys, CryptoKeyReader keyReader,
+            MessageMetadata.Builder msgMetadata, ByteBuf payload) throws PulsarClientException {
 
         if (encKeys.isEmpty()) {
             return payload;
         }
 
         // Update message metadata with encrypted data key
-        encKeys.forEach(keyName -> {
-            if (encryptedDataKeyMap.get(keyName) != null) {
-                msgMetadata.addEncryptionKeys(KeyByteValue.newBuilder().setKey(keyName)
-                    .setValue(ByteString.copyFrom(encryptedDataKeyMap.get(keyName))).build());
+        List<String> keyNameList = encKeys.values();
+        for (int i = 0; i < keyNameList.size(); i++) {
+            String keyName = keyNameList.get(i);
+            if (encryptedDataKeyMap.get(keyName) == null) {
+                // Attempt to load the key. This will allow us to load keys as soon as
+                // a new key is added to producer config
+                addPublicKeyCipher(keyName, keyReader);
+            }
+            Pair<byte[], List<KeyValue>> keyInfo = encryptedDataKeyMap.get(keyName);
+            if (keyInfo != null) {
+                if (keyInfo.getRight().isEmpty()) {
+                    msgMetadata.addEncryptionKeys(EncryptionKeys.newBuilder().setKey(keyName)
+                            .setValue(ByteString.copyFrom(keyInfo.getLeft())).build());
+                } else {
+                    msgMetadata.addEncryptionKeys(EncryptionKeys.newBuilder().setKey(keyName)
+                            .setValue(ByteString.copyFrom(keyInfo.getLeft())).addAllMetadata(keyInfo.getRight()).build());
+                }
             } else {
-                log.warn("Failed to find encrypted Data key for key {}.", keyName);
+                // We should never reach here.
+                log.error("Failed to find encrypted Data key for key {}.", keyName);
             }
-        });
+
+        }
```

Comment:
can we remove `//TODO` at line 393 if it's already implemented?

Reply:
This part is not done yet. Not critical for initial release. But we should revisit.